onnxruntime/core/providers/opencl/opencl_utils.h

```diff
@@ -0,0 +1,412 @@
+    ORT_ENFORCE(ret == CL_MEM_OBJECT_IMAGE2D, mem, " is not Image2D");                                              \
+  }
+
+#define CL_BUFFER_FROM_TENSOR(TENSOR) [&]() {                                                             \
+  cl_mem ptr = const_cast<cl_mem>(static_cast<const std::remove_pointer_t<cl_mem>*>((TENSOR).DataRaw())); \
+  if (ptr) {                                                                                              \
+    CL_CHECK_MEM_OBJECT_IS_BUFFER(ptr);                                                                   \
+  }                                                                                                       \
+  return ptr;                                                                                             \
+}()
+
+#define CL_IMAGE2D_FROM_TENSOR(TENSOR) [&]() {                                                            \
+  cl_mem ptr = const_cast<cl_mem>(static_cast<const std::remove_pointer_t<cl_mem>*>((TENSOR).DataRaw())); \
+  if (ptr) {                                                                                              \
+    CL_CHECK_MEM_OBJECT_IS_IMAGE_2D(ptr);                                                                 \
+  }                                                                                                       \
+  return ptr;                                                                                             \
+}()
+#else
+#define CL_CHECK_MEM_OBJECT_IS_BUFFER(mem)
+#define CL_CHECK_MEM_OBJECT_IS_IMAGE_2D(mem)
+#define CL_BUFFER_FROM_TENSOR(TENSOR) const_cast<cl_mem>(static_cast<const std::remove_pointer_t<cl_mem>*>((TENSOR).DataRaw()))
+#define CL_IMAGE2D_FROM_TENSOR(TENSOR) const_cast<cl_mem>(static_cast<const std::remove_pointer_t<cl_mem>*>((TENSOR).DataRaw()))
+#endif
+
+#define VLOG_CL_NODE()                                          \
+  VLOGS_DEFAULT(0) << "[CL] Node: " << context->GetNodeName()   \
+                   << ", num inputs: " << context->InputCount() \
+                   << ", num outputs: " << context->OutputCount()
+#define VLOG_CL_BUFFER(desc, tensor_ptr)                                      \
+  VLOGS_DEFAULT(0) << "[CL]  " << std::setfill(' ') << std::setw(9) << (desc) \
+                   << " shape " << (tensor_ptr)->Shape()                      \
+                   << "Buffer(" << CL_BUFFER_FROM_TENSOR(*(tensor_ptr)) << ")"
+#define VLOG_CL_IMAGE2D(desc, tensor_ptr)                                     \
+  VLOGS_DEFAULT(0) << "[CL]  " << std::setfill(' ') << std::setw(9) << (desc) \
+                   << " shape " << (tensor_ptr)->Shape()                      \
+                   << "Image2D(" << CL_IMAGE2D_FROM_TENSOR(*(tensor_ptr)) << ")"
+
+namespace onnxruntime {
+namespace opencl {
+
+class NDRange {
+  uint8_t size;
+  size_t values[3];
+
+ public:
+  NDRange() : size(0), values{0, 0, 0} {}
```

Comment:
Can we just use absl::InlinedVector instead of having separate size and values members? see TensorShape implementation for example.

Reply:
- [ ] Use `absl::InlinedVector` for NDRange

Reply:
Done